File: backend/outcome_engine.py
import json
import os
import random
import time
from typing import List, Dict, Tuple, Any, Optional
from models import WinningLine

class OutcomeEngine:
    def __init__(self):
        self.config = {}
        self.buckets = {}
        self.reels = []
        self.symbols = {}
        self.pay_table = {}
        self.lines = {}
        self.is_ready = False
        self.load_config()

    def load_config(self):
        config_path = os.path.join(os.path.dirname(__file__), "game_config_v2.json")
        if not os.path.exists(config_path):
            print(f"Config not found at {config_path}")
            return

        with open(config_path, "r", encoding="utf-8") as f:
            self.config = json.load(f)

        self.reels = self.config["reel_sets"]
        self.symbols = self.config["symbols"]
        self.pay_table = self.config["pay_table"]
        self.lines = self.config["lines"]
        self.buckets_config = self.config["buckets"]
        
        # Normalize buckets config (handle min/max vs min_win/max_win)
        for key, cfg in self.buckets_config.items():
            if "min" in cfg and "min_win" not in cfg:
                cfg["min_win"] = cfg["min"]
            if "max" in cfg and "max_win" not in cfg:
                cfg["max_win"] = cfg["max"]
            # Ensure defaults
            if "min_win" not in cfg: cfg["min_win"] = 0
            if "max_win" not in cfg: cfg["max_win"] = 0

        self.settings = self.config["settings"]
        
        # Initialize buckets
        for key in self.buckets_config:
            self.buckets[key] = []

        print("Config loaded. Initializing buckets...")
        self.initialize_buckets()

    def initialize_buckets(self):
        # Traverse all reel positions (or sample if too large)
        # Reel length is 16. 16^5 = 1,048,576. Feasible.
        
        reel_len = self.config["reels_length"]
        
        # Optimization: If space is too big, use sampling.
        total_combinations = reel_len ** 5
        use_sampling = total_combinations > 2000000 # Limit to ~2M
        
        start_time = time.time()
        
        if use_sampling:
            print(f"State space {total_combinations} too large, using sampling (100k samples).")
            for _ in range(100000):
                stops = [random.randint(0, reel_len - 1) for _ in range(5)]
                self._process_stop(stops)
        else:
            print(f"Traversing all {total_combinations} combinations...")
            # Recursive or Iterative traversal
            # Using Iterative for 5 reels
            import itertools
            ranges = [range(reel_len) for _ in range(5)]
            for stops in itertools.product(*ranges):
                self._process_stop(list(stops))
                
        print(f"Buckets initialized in {time.time() - start_time:.2f}s")
        for k, v in self.buckets.items():
            print(f"Bucket {k}: {len(v)} outcomes")
            
        self.is_ready = True

    def _process_stop(self, stops: List[int]):
        # 1. Build Matrix
        matrix = self._get_matrix_from_stops(stops)
        
        # 2. Calculate Win
        total_win_multiplier, _, is_near_miss = self._calculate_win(matrix)
        
        # 3. Classify
        bucket_name = self._classify_win(total_win_multiplier, is_near_miss)
        
        if bucket_name:
            # Store only stops to save memory
            # We might want to limit the size of buckets if they get too big (e.g. Loss_Random)
            if len(self.buckets[bucket_name]) < 50000: # Cap at 50k per bucket to save RAM
                self.buckets[bucket_name].append(stops)

    def _get_matrix_from_stops(self, stops: List[int]) -> List[List[str]]:
        matrix = []
        reel_len = self.config["reels_length"]
        for r in range(3): # 3 rows
            row = []
            for c in range(5): # 5 cols
                # Reel strip is circular
                idx = (stops[c] + r) % reel_len
                
                # Safety check for reel index
                if c < len(self.reels) and idx < len(self.reels[c]):
                    symbol_id = self.reels[c][idx]
                else:
                    # Fallback if config is inconsistent
                    symbol_id = "L1" 
                    
                row.append(symbol_id)
            matrix.append(row)
        return matrix

    def _calculate_win(self, matrix: List[List[str]]) -> Tuple[float, List[WinningLine], bool]:
        total_payout = 0.0
        winning_lines = []
        
        # Check Lines
        for line_id_str, coords in self.lines.items():
            line_symbols = [matrix[r][c] for r, c in coords]
            count, symbol_id = self._check_line_match(line_symbols)
            
            if count >= 3:
                # Lookup pay
                if symbol_id in self.pay_table:
                    pay_info = self.pay_table[symbol_id]
                    if str(count) in pay_info:
                        multiplier = pay_info[str(count)]
                        total_payout += multiplier
                        
                        symbol_name = "Unknown"
                        if symbol_id in self.symbols:
                            symbol_name = self.symbols[symbol_id]["name"]
                            
                        winning_lines.append(WinningLine(
                            line_id=int(line_id_str),
                            amount=multiplier,
                            symbol=symbol_name,
                            count=count
                        ))
        
        # Check Near Miss (Simplified: 2 Scatters or 4-in-a-row line miss)
        is_near_miss = False
        # Scatter check
        scatter_count = sum(row.count("SCATTER") for row in matrix)
        if scatter_count == 2:
            is_near_miss = True
            
        return total_payout, winning_lines, is_near_miss

    def _check_line_match(self, line: List[str]) -> Tuple[int, str]:
        if not line: return 0, ""
        
        first = line[0]
        match_id = first
        
        # Handle Wilds (WILD is ID "WILD")
        if first == "WILD":
            # Find first non-wild
            for s in line:
                if s != "WILD":
                    match_id = s
                    break
            # If all wilds, match_id remains WILD
        
        count = 0
        for s in line:
            if s == match_id or s == "WILD":
                count += 1
            else:
                break
                
        return count, match_id

    def _classify_win(self, multiplier: float, is_near_miss: bool) -> str:
        if multiplier == 0:
            if is_near_miss:
                return "Loss_NearMiss"
            return "Loss_Random"
        
        # Check tiers
        for tier, cfg in self.buckets_config.items():
            if tier.startswith("Win_Tier"):
                if cfg["min_win"] <= multiplier < cfg["max_win"]:
                    return tier
                # Handle the last tier (max_win might be inclusive or just high)
                if cfg["max_win"] >= 1000 and multiplier >= cfg["min_win"]:
                     return tier
                     
        return "Win_Tier_1" # Fallback

    def spin(self, user_state: Dict[str, Any]) -> Dict[str, Any]:
        if not self.is_ready:
            return {"error": "Engine not ready"}

        bet = user_state.get("current_bet", 10.0)
        balance = user_state.get("wallet_balance", 1000.0)
        initial_balance = user_state.get("initial_balance", 1000.0)
        total_spins = user_state.get("total_spins", 0)
        fail_streak = user_state.get("fail_streak", 0)
        max_historical_balance = user_state.get("max_historical_balance", balance)
        ignore_safety = user_state.get("simulation_mode", False)
        
        # 1. Select Bucket
        bucket_name = self._select_bucket(
            bet, balance, initial_balance, 
            total_spins, fail_streak, 
            ignore_safety=ignore_safety,
            max_historical_balance=max_historical_balance
        )
        if not ignore_safety:
            print(f"[OutcomeEngine] Bet: {bet}, Balance: {balance}, Spins: {total_spins}, FailStreak: {fail_streak}. Selected Bucket: {bucket_name}")
        
        # 2. Pick Outcome from Bucket
        if not self.buckets[bucket_name]:
            # Fallback if bucket empty
            print(f"[OutcomeEngine] Bucket {bucket_name} empty! Fallback to Loss_Random")
            bucket_name = "Loss_Random"
            
        stops = random.choice(self.buckets[bucket_name])

        
        # 3. Generate Details
        matrix = self._get_matrix_from_stops(stops)
        multiplier, winning_lines, _ = self._calculate_win(matrix)
        
        total_payout = multiplier * bet
        
        # Update winning lines with actual amounts
        for wl in winning_lines:
            wl.amount = wl.amount * bet
            
        # Update fail_streak
        new_fail_streak = 0 if total_payout > 0 else fail_streak + 1
            
        return {
            "matrix": matrix,
            "winning_lines": winning_lines,
            "total_payout": total_payout,
            "is_win": total_payout > 0,
            "bucket_type": bucket_name,
            "balance_update": total_payout - bet,
            "fail_streak": new_fail_streak
        }

    def _select_bucket(self, bet: float, balance: float, initial_balance: float, total_spins: int = 0, fail_streak: int = 0, ignore_safety: bool = False, max_historical_balance: float = 0) -> str:
        # 1. PRD Logic: Determine if this spin should be a WIN or LOSS
        base_c = self.settings.get("base_c_value", 0.05)
        win_prob = base_c * (fail_streak + 1)
        
        # Safety: Cap win_prob at 1.0
        if win_prob > 1.0: win_prob = 1.0
        
        is_prd_win = random.random() < win_prob
        
        # 2. Filter available buckets
        weights = {k: v["weight"] for k, v in self.buckets_config.items()}
        
        # If PRD says LOSS, only allow Loss buckets
        if not is_prd_win:
            for k in list(weights.keys()):
                if k.startswith("Win_Tier"):
                    weights[k] = 0
        else:
            # If PRD says WIN, only allow Win buckets
            for k in list(weights.keys()):
                if k.startswith("Loss_"):
                    weights[k] = 0
        
        # 3. Progress Tiers Check
        progress_tiers = self.settings.get("progress_tiers", [])
        if progress_tiers:
            current_tier = None
            # Find the highest tier that matches total_spins
            for tier in sorted(progress_tiers, key=lambda x: x["min_spins"]):
                if total_spins >= tier["min_spins"]:
                    current_tier = tier
                else:
                    break
            
            if current_tier:
                allowed = current_tier.get("allowed_buckets", ["ALL"])
                if "ALL" not in allowed:
                    for k in list(weights.keys()):
                        if k not in allowed:
                            weights[k] = 0

        # 4. High Roller Check
        high_roller_threshold = self.settings.get("high_roller_threshold", 50.0)
        if bet < high_roller_threshold:
            # Remove high tiers
            if "Win_Tier_4" in weights: weights["Win_Tier_4"] = 0
            if "Win_Tier_5" in weights: weights["Win_Tier_5"] = 0
            
        # 5. RTP Safety Cap (Ceiling)
        # Logic: Strictly enforce the ceiling based on initial balance.
        # Simulation and real spins now share the exact same logic.
        
        max_win_ratio = self.settings.get("max_win_ratio", 1.2)
        max_allowed_balance = initial_balance * max_win_ratio
        
        # Normalize weights
        total_weight = sum(weights.values())
        
        # Fallback: If PRD said WIN but all win buckets are filtered out, fallback to Loss
        if total_weight == 0:
            if is_prd_win:
                weights = {k: v["weight"] for k, v in self.buckets_config.items() if k.startswith("Loss_")}
                total_weight = sum(weights.values())
                if total_weight == 0: return "Loss_Random"
            else:
                return "Loss_Random"
        
        # 6. Selection Loop (Try up to 3 times)
        for _ in range(3):
            r = random.uniform(0, total_weight)
            current = 0
            selected = "Loss_Random"
            for k, w in weights.items():
                current += w
                if r <= current:
                    selected = k
                    break
            
            # Validate Safety Cap - ALWAYS enforced now, even in simulation
            max_potential_win = self.buckets_config[selected]["max_win"] * bet
            if balance + max_potential_win > max_allowed_balance:
                # Too risky, try lower tier
                total_weight -= weights[selected]
                weights[selected] = 0
                if total_weight <= 0: 
                    return "Loss_Random"
                continue
            
            return selected
                
        return "Loss_Random"

# Singleton instance
engine = OutcomeEngine()

File: AI-Driven-Slot-Game-JS/backend/src/outcomeEngine.js
const fs = require('fs');
const path = require('path');

// 游戏结果引擎类：负责生成拉霸游戏结果，基于 PRD 算法和桶机制
class OutcomeEngine {
    constructor(configOverride = null) {
        // 初始化配置和数据结构
        this.config = {}; // 游戏配置
        this.buckets = {}; // 桶集合，每个桶存储可能的停止位置
        this.reels = []; // 转轮配置
        this.symbols = {}; // 符号定义
        this.payTable = {}; // 赔付表
        this.lines = {}; // 中奖线定义
        this.isReady = false; // 引擎是否已初始化
        this.bucketStats = {}; // 桶统计信息

        // 如果提供配置覆盖，则使用它，否则加载默认配置
        if (configOverride) {
            this.config = configOverride;
            this._parseConfig();
        } else {
            this.loadConfig();
        }
    }

    // 从文件加载默认配置
    loadConfig() {
        const configPath = path.join(__dirname, 'game_config.json');
        if (!fs.existsSync(configPath)) {
            console.error(`Config not found at ${configPath}`);
            return;
        }
        const rawData = fs.readFileSync(configPath);
        this.config = JSON.parse(rawData);
        this._parseConfig();
    }

    // 解析配置并初始化数据结构
    _parseConfig() {
        // 提取配置中的关键数据
        this.reels = this.config.reel_sets; // 转轮符号集合
        this.symbols = this.config.symbols; // 符号名称映射
        this.payTable = this.config.pay_table; // 赔付倍数表
        this.lines = this.config.lines; // 中奖线坐标
        this.bucketsConfig = this.config.buckets; // 桶配置（胜负分类）
        this.settings = this.config.settings; // 全局设置（RTP、C值等）

        // 标准化桶配置，确保 min_win 和 max_win 字段存在
        for (const key in this.bucketsConfig) {
            const cfg = this.bucketsConfig[key];
            if (cfg.min !== undefined && cfg.min_win === undefined) cfg.min_win = cfg.min;
            if (cfg.max !== undefined && cfg.max_win === undefined) cfg.max_win = cfg.max;
            if (cfg.min_win === undefined) cfg.min_win = 0;
            if (cfg.max_win === undefined) cfg.max_win = 0;
        }

        // 初始化桶数组
        for (const key in this.bucketsConfig) {
            this.buckets[key] = [];
        }

        console.log("Initializing buckets...");
        this.initializeBuckets(); // 开始初始化桶
    }

    // 初始化桶：遍历所有可能的转轮停止位置，分类存储到桶中
    initializeBuckets() {
        const reelLen = this.config.reels_length; // 每个转轮的符号数量
        const totalCombinations = Math.pow(reelLen, 5); // 5个转轮的总组合数
        const useSampling = totalCombinations > 2000000; // 如果组合数过大，使用采样

        const startTime = Date.now();

        if (useSampling) {
            // 使用采样：随机生成10万样本，避免遍历所有组合
            console.log(`State space ${totalCombinations} too large, using sampling (100k samples).`);
            for (let i = 0; i < 100000; i++) {
                const stops = Array.from({ length: 5 }, () => Math.floor(Math.random() * reelLen));
                this._processStop(stops); // 处理每个停止位置
            }
        } else {
            // 遍历所有组合：迭代生成所有可能的停止位置
            console.log(`Traversing all ${totalCombinations} combinations...`);
            const stops = [0, 0, 0, 0, 0];
            while (true) {
                this._processStop([...stops]);

                // 递增停止位置（类似计数器）
                let i = 4;
                while (i >= 0) {
                    stops[i]++;
                    if (stops[i] < reelLen) {
                        break;
                    }
                    stops[i] = 0;
                    i--;
                }
                if (i < 0) break; // 所有组合已遍历
            }
        }

        console.log(`Buckets initialized in ${(Date.now() - startTime) / 1000}s`);
        // 输出每个桶的大小
        for (const key in this.buckets) {
            console.log(`Bucket ${key}: ${this.buckets[key].length} outcomes`);
        }
        this.isReady = true; // 初始化完成
    }

    // 处理单个停止位置：计算中奖并分类到桶中
    _processStop(stops) {
        const matrix = this._getMatrixFromStops(stops); // 生成结果矩阵
        const { totalPayout, isNearMiss } = this._calculateWin(matrix); // 计算中奖
        const bucketName = this._classifyWin(totalPayout, isNearMiss); // 分类桶

        // 将停止位置添加到对应桶中（限制每个桶最多5万条）
        if (bucketName && this.buckets[bucketName]) {
            if (this.buckets[bucketName].length < 50000) {
                this.buckets[bucketName].push(stops);
            }
        }
    }

    // 根据停止位置生成3x5的结果矩阵
    _getMatrixFromStops(stops) {
        const matrix = [];
        const reelLen = this.config.reels_length;
        for (let r = 0; r < 3; r++) { // 3行
            const row = [];
            for (let c = 0; c < 5; c++) { // 5列
                const idx = (stops[c] + r) % reelLen; // 计算符号索引（考虑偏移）
                let symbolId = "L1"; // 默认符号
                if (c < this.reels.length && idx < this.reels[c].length) {
                    symbolId = this.reels[c][idx]; // 从转轮配置获取符号
                }
                row.push(symbolId);
            }
            matrix.push(row);
        }
        return matrix;
    }

    // 计算矩阵的中奖情况：遍历所有中奖线，计算总赔付
    _calculateWin(matrix) {
        let totalPayout = 0.0; // 总赔付倍数
        const winningLines = []; // 中奖线列表

        // 遍历每条中奖线
        for (const lineIdStr in this.lines) {
            const coords = this.lines[lineIdStr]; // 线的坐标 [(r,c), ...]
            const lineSymbols = coords.map(([r, c]) => matrix[r][c]); // 提取线上的符号
            const { count, symbolId } = this._checkLineMatch(lineSymbols); // 检查匹配

            if (count >= 3) { // 至少3个连续匹配
                if (this.payTable[symbolId] && this.payTable[symbolId][count]) {
                    const multiplier = this.payTable[symbolId][count]; // 获取赔付倍数
                    totalPayout += multiplier;
                    
                    let symbolName = "Unknown";
                    if (this.symbols[symbolId]) {
                        symbolName = this.symbols[symbolId].name; // 获取符号名称
                    }

                    winningLines.push({
                        line_id: parseInt(lineIdStr),
                        amount: multiplier, // 倍数（后续会乘以下注）
                        symbol: symbolName,
                        count: count
                    });
                }
            }
        }

        // 检查近失（Near Miss）：如果有2个SCATTER，标记为近失
        let isNearMiss = false;
        let scatterCount = 0;
        for (const row of matrix) {
            for (const s of row) {
                if (s === "SCATTER") scatterCount++;
            }
        }
        if (scatterCount === 2) isNearMiss = true;

        return { totalPayout, winningLines, isNearMiss };
    }

    // 检查线上的符号匹配：计算连续相同符号的数量
    _checkLineMatch(line) {
        if (!line || line.length === 0) return { count: 0, symbolId: "" };

        const first = line[0];
        let matchId = first;

        // 如果第一个是WILD，寻找下一个非WILD符号作为匹配目标
        if (first === "WILD") {
            for (const s of line) {
                if (s !== "WILD") {
                    matchId = s;
                    break;
                }
            }
        }

        let count = 0;
        for (const s of line) {
            if (s === matchId || s === "WILD") { // WILD可以匹配任何符号
                count++;
            } else {
                break; // 连续匹配中断
            }
        }
        return { count, symbolId: matchId };
    }

    // 根据赔付倍数和近失情况分类桶
    _classifyWin(multiplier, isNearMiss) {
        if (multiplier === 0) {
            return isNearMiss ? "Loss_NearMiss" : "Loss_Random"; // 近失或普通亏损
        }

        // 根据倍数范围匹配胜级桶
        for (const tier in this.bucketsConfig) {
            if (tier.startsWith("Win_Tier")) {
                const cfg = this.bucketsConfig[tier];
                if (multiplier >= cfg.min_win && multiplier < cfg.max_win) {
                    return tier;
                }
                if (cfg.max_win >= 1000 && multiplier >= cfg.min_win) { // 处理大奖
                    return tier;
                }
            }
        }
        return "Win_Tier_1"; // 默认胜级
    }

    // 执行一次拉霸旋转：选择桶、生成结果、计算赔付
    spin(userState, runtimeConfig = null) {
        if (!this.isReady) return { error: "Engine not ready" };

        // 提取用户状态参数
        const bet = userState.current_bet || 10.0; // 下注金额
        const balance = userState.wallet_balance || 1000.0; // 当前余额
        const initialBalance = userState.initial_balance || 1000.0; // 初始余额
        const totalSpins = userState.total_spins || 0; // 总旋转次数
        const failStreak = userState.fail_streak || 0; // 连败次数
        const maxHistoricalBalance = userState.max_historical_balance || balance; // 历史最高余额
        const historicalRtp = userState.historical_rtp || 0.0; // 历史 RTP
        const ignoreSafety = userState.simulation_mode || false; // 是否忽略安全检查（模拟模式）

        // 选择桶（基于 PRD 算法和用户状态）
        let bucketName = this._selectBucket(
            bet, balance, initialBalance, totalSpins, failStreak,
            ignoreSafety, maxHistoricalBalance, historicalRtp, runtimeConfig
        );

        if (!ignoreSafety) {
            console.log(`[OutcomeEngine] Bet: ${bet}, Balance: ${balance}, Spins: ${totalSpins}, FailStreak: ${failStreak}. Selected Bucket: ${bucketName}`);
        }

        // 如果桶为空，回退到 Loss_Random
        if (!this.buckets[bucketName] || this.buckets[bucketName].length === 0) {
            console.log(`[OutcomeEngine] Bucket ${bucketName} empty! Fallback to Loss_Random`);
            bucketName = "Loss_Random";
        }

        // 从桶中随机选择一个停止位置
        const stops = this.buckets[bucketName][Math.floor(Math.random() * this.buckets[bucketName].length)];
        const matrix = this._getMatrixFromStops(stops); // 生成矩阵
        const { totalPayout, winningLines } = this._calculateWin(matrix); // 计算中奖

        const totalPayoutAmount = totalPayout * bet; // 实际赔付金额
        
        // 更新中奖线金额（乘以下注）
        winningLines.forEach(wl => wl.amount = wl.amount * bet);

        // 更新连败次数
        const newFailStreak = totalPayoutAmount > 0 ? 0 : failStreak + 1;

        return {
            matrix, // 结果矩阵
            winning_lines: winningLines, // 中奖线详情
            total_payout: totalPayoutAmount, // 总赔付金额
            is_win: totalPayoutAmount > 0, // 是否中奖
            bucket_type: bucketName, // 使用的桶类型
            balance_update: totalPayoutAmount - bet, // 余额变化
            fail_streak: newFailStreak // 新连败次数
        };
    }

    // 选择桶：基于 PRD 算法和多种因素选择合适的桶
    _selectBucket(bet, balance, initialBalance, totalSpins, failStreak, ignoreSafety, maxHistoricalBalance, historicalRtp, runtimeConfig) {
        // 使用运行时配置或默认设置
        let settings = this.settings;
        let bucketsConfig = this.bucketsConfig;

        if (runtimeConfig) {
            settings = runtimeConfig.settings || this.settings;
            const rawBuckets = runtimeConfig.buckets || this.bucketsConfig;
            bucketsConfig = {};
            for (const k in rawBuckets) {
                const cfg = { ...rawBuckets[k] };
                if (cfg.min !== undefined && cfg.min_win === undefined) cfg.min_win = cfg.min;
                if (cfg.max !== undefined && cfg.max_win === undefined) cfg.max_win = cfg.max;
                if (cfg.min_win === undefined) cfg.min_win = 0;
                if (cfg.max_win === undefined) cfg.max_win = 0;
                bucketsConfig[k] = cfg;
            }
        }

        // 基础 C 值（PRD 概率常数）
        let baseC = settings.base_c_value || 0.05;
        const targetRtp = settings.target_rtp || 0.97;

        // 根据历史 RTP 动态调整 C 值（RTP 调控）
        if (totalSpins > 50) {
            const rtpRatio = targetRtp > 0 ? historicalRtp / targetRtp : 1.0;
            if (rtpRatio < 0.5) baseC *= 2.5; // RTP 太低，增加中奖概率
            else if (rtpRatio < 0.7) baseC *= 1.8;
            else if (rtpRatio < 0.8) baseC *= 1.2;
            else if (rtpRatio < 0.95) baseC *= 1.1;
            else if (rtpRatio > 2.0) baseC *= 0.3; // RTP 太高，减少中奖概率
            else if (rtpRatio > 1.5) baseC *= 0.5;
            else if (rtpRatio > 1.05) baseC *= 0.6;
        }

        // 计算中奖概率（PRD 公式）
        let winProb = baseC * (failStreak + 1);
        if (winProb > 1.0) winProb = 1.0;

        // 决定是否中奖
        const isPrdWin = Math.random() < winProb;

        // 初始化桶权重
        const weights = {};
        for (const k in bucketsConfig) {
            weights[k] = bucketsConfig[k].weight;
        }

        // 根据 PRD 结果过滤桶
        if (!isPrdWin) {
            // 不中奖：禁用所有胜级桶
            for (const k in weights) {
                if (k.startsWith("Win_Tier")) weights[k] = 0;
            }
        } else {
            // 中奖：禁用所有亏损桶
            for (const k in weights) {
                if (k.startsWith("Loss_")) weights[k] = 0;
            }
        }

        // 进度阶梯：根据总旋转次数限制可用桶
        const progressTiers = settings.progress_tiers || [];
        if (progressTiers.length > 0) {
            let currentTier = null;
            const sortedTiers = [...progressTiers].sort((a, b) => a.min_spins - b.min_spins);
            for (const tier of sortedTiers) {
                if (totalSpins >= tier.min_spins) {
                    currentTier = tier;
                } else {
                    break;
                }
            }

            if (currentTier) {
                const allowed = currentTier.allowed_buckets || ["ALL"];
                if (!allowed.includes("ALL")) {
                    for (const k in weights) {
                        if (!allowed.includes(k)) weights[k] = 0;
                    }
                }
            }
        }

        // 高额玩家过滤：小额下注禁用大奖桶
        const highRollerThreshold = settings.high_roller_threshold || 50.0;
        if (bet < highRollerThreshold) {
            if (weights["Win_Tier_4"]) weights["Win_Tier_4"] = 0;
            if (weights["Win_Tier_5"]) weights["Win_Tier_5"] = 0;
        }

        // 最大余额限制：防止单次中奖超过初始余额的倍数
        const maxWinRatio = settings.max_win_ratio || 1.2;
        const maxAllowedBalance = initialBalance * maxWinRatio;

        let totalWeight = Object.values(weights).reduce((a, b) => a + b, 0);

        // 如果权重为0，尝试回退
        if (totalWeight === 0) {
            if (isPrdWin) {
                // 中奖但无胜级桶可用，回退到亏损桶
                for (const k in bucketsConfig) {
                    if (k.startsWith("Loss_")) weights[k] = bucketsConfig[k].weight;
                }
                totalWeight = Object.values(weights).reduce((a, b) => a + b, 0);
                if (totalWeight === 0) return "Loss_Random";
            } else {
                return "Loss_Random";
            }
        }

        // 加权随机选择桶，最多重试3次（避免超出余额限制）
        for (let i = 0; i < 3; i++) {
            const r = Math.random() * totalWeight;
            let current = 0;
            let selected = "Loss_Random";
            
            for (const k in weights) {
                current += weights[k];
                if (r <= current) {
                    selected = k;
                    break;
                }
            }

            // 检查是否会超出最大余额
            const maxPotentialWin = bucketsConfig[selected].max_win * bet;
            if (balance + maxPotentialWin > maxAllowedBalance) {
                // 移除此桶，重新选择
                totalWeight -= weights[selected];
                weights[selected] = 0;
                if (totalWeight <= 0) return "Loss_Random";
                continue;
            }
            return selected;
        }

        return "Loss_Random"; // 默认回退
    }
}

module.exports = OutcomeEngine;
